drivers/gpio/gpio_hogs.c

```diff
@@ -0,0 +1,126 @@
+/*
+ * Copyright (c) 2022-2023 Vestas Wind Systems A/S
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/devicetree.h>
+#include <zephyr/drivers/gpio.h>
+#include <zephyr/init.h>
+#include <zephyr/logging/log.h>
+
+LOG_MODULE_REGISTER(gpio_hogs, CONFIG_GPIO_LOG_LEVEL);
+
+struct gpio_hog_dt_spec {
+	gpio_pin_t pin;
+	gpio_flags_t flags;
+};
+
+struct gpio_hogs {
+	const struct device *port;
+	const struct gpio_hog_dt_spec *specs;
+	size_t num_specs;
+};
+
+/* Static initializer for a struct gpio_hog_dt_spec */
+#define GPIO_HOG_DT_SPEC_GET_BY_IDX(node_id, idx)						\
+	{											\
+		.pin = DT_GPIO_HOG_PIN_BY_IDX(node_id, idx),					\
+		.flags = DT_GPIO_HOG_FLAGS_BY_IDX(node_id, idx) |				\
+			 COND_CODE_1(DT_PROP(node_id, input), (GPIO_INPUT),			\
+				     COND_CODE_1(DT_PROP(node_id, output_low),			\
+						 (GPIO_OUTPUT_INACTIVE),			\
+						 COND_CODE_1(DT_PROP(node_id, output_high),	\
+							     (GPIO_OUTPUT_ACTIVE), (0)))),	\
+	}
```

Comment:
This can be written more concisely and with less nesting as:
```c
#define GPIO_HOG_DT_SPEC_GET_BY_IDX(node_id, idx)                                       \
        {                                                                               \
                .pin = DT_GPIO_HOG_PIN_BY_IDX(node_id, idx),                            \
                .flags = DT_GPIO_HOG_FLAGS_BY_IDX(node_id, idx) |                       \
                         (DT_PROP(node_id, input) ? GPIO_INPUT : 0) |                   \
                         (DT_PROP(node_id, output_low) ? GPIO_OUTPUT_INACTIVE : 0) |    \
                         (DT_PROP(node_id, output_high) ? GPIO_OUTPUT_ACTIVE : 0),      \
        }
```
There is another improvement too.  If multiple flags from input, output-low and output-high are used it will construct an invalid gpio flags bitmask, which will then cause a runtime assert.

The existing code will just pick one of the flags to use, ignore the rest, and silently do what is possibly the wrong thing with no errors or warnings.

Reply:
The `input` property takes precedence over the `output-low` property which takes precedence over the `output-high` property. This is documented in the DTS binding and consistent with how the Linux kernel handles these properties.

Reply:
It's inconsistent with Zephyr's existing gpio API, which considers it an error to pass conflicting modes in `gpio_flags_t`.

It also seems highly unlikely to me that anyone would ever intentionally configure a GPIO as both an input and an output, or be both high and low, and desire the priority of one mode over the other to be used.  I think it's safe to say every time this happens, it's a mistake.

Do you really think anyone would do that intentionally?